src/python/pants/engine/target.py

```diff
@@ -163,10 +174,10 @@ def __str__(self) -> str:
 
 @frozen_after_init
 @dataclass(unsafe_hash=True)
-class AsyncField(Field, metaclass=ABCMeta):
-    """A field that needs the engine in order to be hydrated.
+class AsyncField(Field[_DefaultBase], metaclass=ABCMeta):
+    """A field that needs to extract information from an Address via the engine to be hydrated.
```

Comment:
This isn't quite right. You might not need the address at all for hydration, only, it's often useful.

The original description was intentionally vague - why you need the engine for hydration depends on every field. What the hydration looks like also depends on the field. The only thing we can generalize is "you need the engine for some reason to fully get the value you care about".